### PR TITLE
#8 Implemented Abstraction for Logging users

### PR DESCRIPTION
**Pull request by mparq**

### CODE REVIEW

**Comment by calebdre** [comment]
Can you make a new commit copying the functionality to the PatientEngine?

**Comment by calebdre** [comment]
Also, make sure that the LogManagers are the ones that know how to assemble an action string.

**Comment by mparq** [comment]
ChoiceLogManager should be good to go. Right now, it handles logging the users choice action string and sending that to the server. Do you want me to also make LabNotebookLogManager specific as well? Meaning, it will only handle writing those specific logs to the notebook within the simulation. Originally, I made it just a general wrapper into the LabNotebook so the user would still have to build their own string and pass that into the NotebookLogManager. I feel like we wouldn't even need this, since labnotebook provides its own add method. It also doesn't make sense to have a specific LabNotebookLogManager inside PatientEngine, because there are other places where we add to the lab notebook. Within PatientEngine, we are writing specific strings to the lab notebook, but maybe this functionality should be handled within our ChoiceLogManager as well. I can try to make this happen, but it does appear to be a little difficult due to the difference in building loggableString between onBinaryClick and onSubmitClick.

**Comment by mparq** [comment]
Now ChoiceLogManager handles all logging during the simulation. However, some bugs have been introduced. From what I can tell, the code to grab the current Patient if it exists on the page is flawed. I'll have to go back to previous commits and see exactly where, but my suspicion is it happened in the move from PageController to PatientEngine. Another bug I noticed is that popup windows are no longer handled correctly. This may be due to something I changed. 

I also wanted to ask something, and this is in regards to the async api calls we make when logging the action string of a choice. Before, you had the code for changing the page attached as a callback for after the api call completes. I moved the code for changing code to not be attached as a call back since 1) ChoiceLogManager didn't seem like it should be handling page changes and 2) it made the code look a little bit nicer. Do you think this will introduce anything dangerous in the code? If so, we'll want to think of what we can do to make the page change wait until after the api call completes.

**Comment by mparq** [comment]
Merged with latest upstream making use of LocalStorage and page contexts. We still have the problem of logging looking a little off regarding identifying When we should have a message that looks like "Patient:" and one that looks like "Question:". I also commented out the lines in PatientEngine which attempt to identify whether the page is a popup. I get an error in my browser stating Object.values is not a function. Does this work for your browser?

File: scisim/static/js/Utils/LogManager.js [hunk 1]
```diff
@@ -0,0 +1,54 @@
...
+    }else{
+        loggableString += "Question: " + choiceInfo.prev.text() + " You said: ";
+    }
+
+    loggableString += choiceInfo.choice;
+
+    labnotebook.add(loggableString.replace("}", ""));
+
+    if (choiceInfo.page_context.hasOwnProperty("patient")) {
+        choiceInfo.page_context.patient.choices.push(loggableString);
```
Comment (calebdre) [inline scisim/static/js/Utils/LogManager.js:25]: note that when you update a context, you have to call chain.updateContext on the context.

File: scisim/static/js/Engines/Patient/PatientEngine.js [hunk 2]
```diff
@@ -1,11 +1,146 @@
...
+
+PatientEngine.prototype.disableChoices = function() {
+	var choices = $('.choice');
+	for (var i = 0; i < choices.length; i++) {
+		choices.eq(i).addClass("disabled");
+	}
+};
+
+PatientEngine.prototype.changePage = function(destination) {
+	this.choiceLogger.flushLog();
```
Comment (calebdre) [inline scisim/static/js/Engines/Patient/PatientEngine.js:141]: it doesn't look like the new page will be rendered before the api requests finish, however i don't think this is a problem. If it does become one, we can wrap it in a deferred.

**Comment by calebdre** [comment]
LogManager looks pretty good - it cleans up the PatientEngine pretty nicely. I don't think the new page is rendered after the logging api calls are made, but i don't think this is an issue.  

For logging, we just want to append the patient name before the log if the name is available. I used the patient object in places however the initial detecting a patient doesn't seem to be working - i'll have to look into that later today. Right now, the logging doesn't have a bug but the PatientManager does.  

Object.Values should be working..i noticed you tried it before you pulled the latest version - have you tried it with the new changes? It works on my browser (Firefox.) I added it to the polyfils/Object because it's currently not a native Javascript method

**Comment by mparq** [comment]
I did try it with the latest version I believe. Let me try it one more time to make sure. I also fixed up LogManager to call chain.updateContext and use storageHelper.

**Comment by mparq** [comment]
Yeah, strange. I see the polyfils/Object.js and it is being imported in index.html, but I keep getting that type error Object.values is not a function. I tried in both Chrome and Firefox with the same result.

**Comment by mparq** [comment]
Ah, okay. In my index.html, js/polyfils/Object.js was being imported, but it should have been js/Polyfils/Object.js. I guess I'll also commit that fix and uncomment the lines that were causing issue before. Hopefully this fixes the popup issue I was getting.

**Comment by mparq** [comment]
Yep! Popups are work correctly now.

**Pull request merged**